### PR TITLE
DownloadFile behaves unexpectedly in the case of OperationCanceledExceptions

### PR DESCRIPTION
**Pull request by ChadNedzlek**
We have some builds that are calling the DownloadFile task, and they seem to be
silently misbehaving: no errors, just hangs for a few hours and then we have to kill the process.

Looking at the code, there are a few paths through the code that result in either infinite loops or silently downloading nothing and returning success.

In particular, if a server returns a failed response, like a 500, and then the task is cancelled, it enters an infinite loop, because canRetry isn't refreshed in the loop.

Also, if a request times out the first time, where HttpClient.Send throws an OperationCanceledException, the task returns success without downloading anything.

I added tests for these two cases, and they fail without this change.

Hopefully this change is an acceptable one... I tried hard not to change any existing code behaviors (other than infinite loop and silent failure).  This will still fail and not-retry if any request times out, but that was the existing behavior (it just correctly reports and error and returns false now, instead of not downloading the requested file), and perhaps that behavior isn't desirable, in which case, IsRetriable should have a clause for OperationCanceled that returns true.

### CODE REVIEW

File: src/Tasks/DownloadFile.cs [hunk 1]
```diff
@@ -77,36 +83,46 @@ public override bool Execute()
             }
 
             int retryAttemptCount = 0;
-            bool canRetry = false;
+            
+            CancellationToken cancellationToken = _cancellationTokenSource.Token;
 
-            do
+            while(true)
```
Comment (Forgind) [inline src/Tasks/DownloadFile.cs]: Why did you change the type of loop? I personally would prefer to see a single while or do/while loop without break statements if possible, that is, putting `canRetry = false;` in instead of breaks.

File: src/Tasks/DownloadFile.cs [hunk 2]
```diff
@@ -77,36 +83,46 @@ public override bool Execute()
...
                 try
                 {
-                    Download(uri);
+                    await DownloadAsync(uri, cancellationToken);
                     break;
                 }
-                catch (OperationCanceledException)
+                catch (OperationCanceledException e) when (e.CancellationToken == cancellationToken)
                 {
+                    // This task is being cancelled, exit the loop
```
Comment (Forgind) [inline src/Tasks/DownloadFile.cs]:
```suggestion
                    // This task is being cancelled. Exit the loop.
```

File: src/Tasks.UnitTests/DownloadFile_Tests.cs [hunk 3]
```diff
@@ -245,6 +245,68 @@ public void RetryOnResponseError()
...
+                }),
+                Retries = 2,
+                RetryDelayMilliseconds = 100,
+                SourceUrl = "http://notfound/foo.txt"
+            };
+
+            var runaway = Task.Run(() => downloadFile.Execute());
+            await Task.Delay(TimeSpan.FromSeconds(1));
+            runaway.IsCompleted.ShouldBeTrue("Task did not cancel");
+            if (!runaway.IsCompleted)
```
Comment (Forgind) [inline src/Tasks.UnitTests/DownloadFile_Tests.cs]: Isn't this redundant given the previous line? If not, can you add braces around it or put it on one line?

File: src/Tasks/DownloadFile.cs [hunk 4]
```diff
@@ -69,6 +70,11 @@ public void Cancel()
         }
 
         public override bool Execute()
+        {
+            return ExecuteAsync().GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (ChadNedzlek) [reply src/Tasks/DownloadFile.cs]: This is the (unforunate) sync wrapper around the async code, because, to my knowledge, there is no async entry point into a build Task.  So this is the only place that can't use it.

File: src/Tasks/DownloadFile.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/Tasks/DownloadFile.cs]: That is the source of the bugs.  There are many cases that _should_ break out of the loop, but failed to do so before, because it carried the value from the previous iteration of the loop unexpectedly.  There is no way to scope the variable to the body of loop, so you HAVE to carry the value between iterations, even though each iteration has nothing to do with the previous one.  By using explicit breaks, it's super clear when something means to break out of the loop.

I could change it back, and then just make sure to reset "canRetry" to false at the beginning of every loop... and then make sure it's set to true in every case where we want to retry... but that seems like a complicated and error prone way of saying "break" when we want to break.

File: src/Tasks.UnitTests/DownloadFile_Tests.cs (hunk 3 above)
Comment (ChadNedzlek) [reply src/Tasks.UnitTests/DownloadFile_Tests.cs]: I suppose it is, if the assumption is that a failed "Should" will _always_ abort the test.  The problem is that if this isn't trye, the link 306 will wait infinitely.  I wanted to be extra sure the test didn't enter an infinite loop, since it's really hard to report on those.

I'd rather leave the valve, but I'll add the braces and add a comment explaining why it's there.

**Comment by ChadNedzlek** [comment]
I think I resolved all the outstanding issues and have pushed an updated change merged with master

File: src/Tasks.UnitTests/DownloadFile_Tests.cs (hunk 3 above)
Comment (rainersigwald) [reply src/Tasks.UnitTests/DownloadFile_Tests.cs]: Since xUnit assertions work via exception, I think it's safe to assume that the `ShouldBeTrue` will end the test if it fails. We also have #4973 and https://github.com/dotnet/arcade/pull/3255 coming which should help catch hung tests in the general case.

Comment (ChadNedzlek) [reply src/Tasks.UnitTests/DownloadFile_Tests.cs]: The arcade situation will almost certainly cause it to just... lose the test results, but I'm find removing this test hardening.  Committed an update.
